### PR TITLE
Loading cache improve eviction use policy

### PR DESCRIPTION
**Pull request by vladzcloudius**
This series introduces a new version of a loading_cache class.
The old implementation was susceptible to a "pollution" phenomena when frequently used entry can get evicted by an intensive burst of "used once" entries pushed into the cache.

The new version is going to have a privileged and unprivileged cache sections and there's a new loading_cache template parameter - SectionHitThreshold. The new cache algorithm goes as follows:
  * We define 2 dynamic cache sections which total size should not exceed the maximum cache size.
  * New cache entry is always added to the "unprivileged" section.
  * After a cache entry is read more than SectionHitThreshold times it moves to the second cache section.
  * Both sections' entries obey expiration and reload rules in the same way as before this patch.
  * When cache entries need to be evicted due to a size restriction "unprivileged" section's
    least recently used entries are evicted first.

More details may be found in #8674.

In addition, during a testing another issue was found in the authorized_prepared_statements_cache: #9590.
There is a patch that fixes it as well.

### CODE REVIEW

**Comment by vladzcloudius** [comment]
@nyh FYI.
The series is tested using new tests I added + there is a set of new dtest tests that is going to be sent as soon as this series is merged.

**Comment by vladzcloudius** [comment]
Hold on - there is some issue with a unit test.

**Comment by vladzcloudius** [comment]
> Hold on - there is some issue with a unit test.

Force-pushed the fix.
It's good for merging now, @nyh

File: utils/loading_cache.hh [hunk 1]
```diff
@@ -86,6 +96,7 @@ struct simple_entry_size {
 ///
 /// \tparam Key type of the cache key
 /// \tparam Tp type of the cached value
+/// \tparam SectionHitThreshold number of hits after which a cache item is going to be moved to the privileged cache section.
```
Comment (nyh) [inline utils/loading_cache.hh:105]: I've said this too many times already but the word "after" is ambiguous. It should have been phrased differently.

**Review by nyh — APPROVED** [review]
I'll merge.

**Pull request merged**